plugins/e57/io/E57Reader.hpp

```diff
@@ -1,138 +1,87 @@
-        std::unique_ptr<e57::CompressedVectorReader> m_dataReader;
-        std::shared_ptr<e57::Scan> m_scan;
-    };
+using namespace e57;
+using namespace std;
 
+class PDAL_DLL E57Reader : public Reader, public Streamable
+{
 public:
-    E57Reader()
-    {}
-
-    E57Reader(std::string filename);
-    ~E57Reader();
-    E57Reader(const E57Reader &) = delete;
-    E57Reader& operator=(const E57Reader&) = delete;
-
-    std::string getName() const;
-
-    /// Get the total number of points in the scan
-    point_count_t getNumberPoints() const;
-
-    /// Gets the scan index of a given point index
-    int getScanIndex(point_count_t) const;
-
-    /// Gets the dimensions present within the set of clouds
-    std::set<std::string> getDimensions();
-
-    /// returns the scans
-   std::vector<std::shared_ptr<e57::Scan>> getScans() const;
+    E57Reader();
+    std::string getName() const override;
 
 private:
-
     /* Pdal section */
-    virtual void addDimensions(PointLayoutPtr layout);
-    virtual void initialize();
-    virtual bool processOne(PointRef& point);
-    virtual point_count_t read(PointViewPtr view, point_count_t count);
-    virtual QuickInfo inspect();
-
-    void openFile(const std::string &filename);
-    void closeFile();
-    void setupReader(point_count_t pointNumber);
-    point_count_t extractNumberPoints() const;
-    void extractScans();
-
-    // members
-    std::unique_ptr<e57::ImageFile> m_imf;
-    std::vector<std::shared_ptr<e57::Scan>> m_scans;
-
-    // Allows construction by filename
-    std::string m_filenameManual;
-
-    // E57 dimensions that are common to all scans
-    std::set<std::string> m_validDimensions;
-
-    // Chunk data reader
-    std::unique_ptr<ChunkReader> m_chunk;
-    point_count_t m_currentPoint; // Variable for streaming mode
-
-    // Cache the total number of points
-    point_count_t m_pointCount;
+    virtual void addDimensions(PointLayoutPtr layout) override;
+    virtual void initialize() override;
+    virtual point_count_t read(PointViewPtr view, point_count_t count) override;
+    virtual void done(PointTableRef table) override;
+    virtual bool processOne(PointRef&) override;
+    virtual void ready(PointTableRef&) override;
+
+    // Private members
+    bool fillPoint(PointRef& point);
+    point_count_t readNextBatch();
+    void setupReader();
+
+    std::unique_ptr<ImageFile> m_imf;
+    std::unique_ptr<VectorNode> m_data3D;
+    std::unique_ptr<CompressedVectorReader> m_reader;
+    std::unique_ptr<StructureNode> m_e57PointPrototype;
+
+    std::map<std::string, std::vector<double>> m_doubleBuffers;
+    std::map<pdal::Dimension::Id, double> m_rescaleFactors;
+    vector<SourceDestBuffer> m_destBuffers;
+
+    point_count_t m_currentIndex;
+    point_count_t m_pointsInCurrentBatch;
+    point_count_t m_defaultChunkSize;
+    signed int m_currentScan;
+
+    // Pose information
+    double m_translation[3] = {0};
```

Comment:
I introduced the Scan class because this translation and rotation infos are related to each scan. I think it makes sense to keep it unless it is a performance issue but I don't think it was.